test/helpers/foundry/extract.ts

```diff
@@ -0,0 +1,211 @@
+) {
+  const downloads: {
+    path: string;
+    binary: string;
+    checksum?: string;
+  }[] = [];
+  await pipeline(
+    startDownload(url),
+    extractTar(
+      {
+        cwd: dir,
+        transform: (entry) => {
+          const absolutePath = entry.absolute;
+          if (!absolutePath) {
+            throw new Error('Missing absolute path for entry');
+          }
+
+          if (checksumAlgorithm) {
+            const hash = createHash(checksumAlgorithm);
+            const passThrough = new Minipass({ async: true });
+            passThrough.pipe(hash);
+            passThrough.on('end', () => {
+              downloads.push({
+                path: absolutePath,
+                binary: entry.path,
+                checksum: hash.digest('hex'),
+              });
+            });
+            return passThrough;
+          }
+
+          // When no checksum is needed, record the entry and return undefined
+          // to use the original stream without transformation
+          downloads.push({
+            path: absolutePath,
+            binary: entry.path,
+          });
+          return undefined;
+        },
+      },
+      binaries,
+    ),
+  );
+  return downloads;
+}
+/**
+ * Extracts the binaries from a zip archive.
+ *
+ * @param url - The URL of the archive to extract the binaries from
+ * @param binaries - The list of binaries to extract
+ * @param dir - The destination directory
+ * @param checksumAlgorithm - The checksum algorithm to use
+ * @returns The list of binaries extracted
+ */
+
+async function extractFromZip(
+  url: URL,
+  binaries: string[],
+  dir: string,
+  checksumAlgorithm?: string,
+) {
+  const agent = new (url.protocol === 'http:' ? HttpAgent : HttpsAgent)({
+    keepAlive: true,
+  });
+  const source: Source = {
+    async size() {
+      const download = startDownload(url, { agent, method: 'HEAD' });
+      const response = await download.response();
+      const contentLength = response.headers['content-length'];
+      return contentLength ? parseInt(contentLength, 10) : 0;
+    },
+    stream(offset: number, bytes: number) {
+      const options = {
+        agent,
+        headers: {
+          range: `bytes=${offset}-${bytes ? offset + bytes : ''}`,
+        },
+      };
+      return startDownload(url, options);
+    },
+  };
+
+  const { files } = await Open.custom(source, {});
+  const filtered = files.filter(({ path }) =>
+    binaries.includes(basename(path, extname(path))),
+  );
+  return await Promise.all(
+    filtered.map(async ({ path, stream }) => {
+      const dest = join(dir, path);
+      const entry = stream();
+      const destStream = createWriteStream(dest);
+      if (checksumAlgorithm) {
+        const hash = createHash(checksumAlgorithm);
+        const hashStream = async function* (entryStream: Entry) {
+          for await (const chunk of entryStream) {
+            hash.update(chunk);
+            yield chunk;
+          }
+        };
```

Comment:
Instead of `tee`ing the stream I'm just reading the data as it passes through the `pipeline` and then letting it continue on untouched. This could be done with a `Stream.PassThrough`, but that would requires a creating a new `class` that extends `Stream.PassThrough`.

Reply:
A noop transform would do it